### PR TITLE
Add favorites

### LINKED ISSUE #37: Button to add a entry to favorites
The favorite entries will be always in the list os items.

### LINKED ISSUE #85: [Feature request] Pin value
Hi,
I used clipIt (https://apps.ubuntu.com/cat/applications/precise/clipit/) on Unity, but after I switched to gnome I found this extension.
Only one feature, that I need here, is to store a persistent value into the history list. I have a special character in my name and it's super useful for everyday usage. And I think it would be great for other people too (not only for that name thing).

How it looks like: 
screenshot from 2017-10-23 14-29-46

ö is the static item

### PR DESCRIPTION
**Pull request by freeroot**
· Favorites are not deleted by "clear history"
· A full/empty star to toggle favorites
· One favorite boolean associate to each item
· Registry cache file to save favorites across sessions
· Change "Disable cache file" option into "Cache only favorites"

Enhancement fix #37 and fix #85 .

example

### CODE REVIEW

**Comment by JPvRiel** [comment]
> Change "Disable cache file" option into "Cache only favorites"

Nice, I was thinking that 'disable cache' would break this new feature too, but I see you're a step ahead.

File: extension.js [hunk 1]
```diff
@@ -155,7 +155,7 @@ const ClipboardIndicator = Lang.Class({
 
             // Add cached items
             clipHistory.forEach(function (buffer) {
-                that._addEntry(buffer);
+                that._addEntry(buffer[0], buffer[1]);
```
[the remark this answers is not in the document]
Comment (Tudmotu) [reply extension.js]: On second thought, I can live with the structure of nested arrays for start, and when we feel like it we can improve the structure. 
But still we have to make this backwards compatible.

File: extension.js [hunk 2]
```diff
@@ -336,10 +376,20 @@ const ClipboardIndicator = Lang.Class({
...
+                }
+            } else {
+                return [menuItem.clipContents, menuItem.clipFavorite];
+            }
+        });
 
-        writeRegistry(this.clipItemsRadioGroup.map(function (menuItem) {
-            return menuItem.clipContents;
+        writeRegistry(registry.filter(function (menuItem) {
+            if (menuItem != 'null') {
```
[the remark this answers is not in the document]
Comment (freeroot) [reply extension.js]: Because of a bad using of map just above.
Fix in 0e8b026

File: extension.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (freeroot) [reply extension.js]: I follow your advices in 0e8b026.

**Comment by freeroot** [comment]
I think d62227c is enough for a full change but I'm not expert..

File: utils.js [hunk 3]
```diff
@@ -95,14 +95,7 @@ function readRegistry (callback) {
 
                 if (success) {
                     try {
-                        let max_size = SettingsSchema.get_int(Prefs.Fields.HISTORY_SIZE);
```
Comment (Tudmotu) [inline utils.js]: Hmm, why did you remove this?

Comment (freeroot) [reply utils.js]: To avoid deleting the favourites in the init.

**Comment by danboid** [comment]
This PR is looking promising - good work freeroot - but there are a couple of issues for me under Ubuntu 17.10 amd64 / GNOME 3.26.1:

* The extension works under  GNOME / X.org but GNOME / Wayland fails to start with this extension installed.

* New items added to the clipboard are favourited by default - they shouldn't be.

Thanks

**Comment by freeroot** [comment]
@danboid 
1d4e389 fix a bad merge (the first I did on github^^)

I work in Xorg, I don't believe it couldn't work with Wayland. Do you know how to investigate ?

**Comment by danboid** [comment]
Are you running a flavour of Ubuntu 17.10? I initially installed Ubuntu MATE 17.10 but it was very easy to install GNOME and Wayland alongside MATE / Xorg. The main thing to note is that you must be running gdm3 instead of say lightdm (which is used by Ubuntu MATE) in order to log into a Wayland GNOME session.

I was able to switch from X11 and MATE to GNOME 3 and Wayland simply by running:

sudo apt install gnome-session-wayland gdm3 chrome-gnome-shell

Then logging out and picking GNOME (not GNOME X.org) from the gdm options that show when you've selected your user.

Wayland doesn't have a separate log file for errors - its all rolled into the same logs as dmesg etc ad viewed with journalctl as with anything else under systemd.

**Comment by freeroot** [comment]
I'm on Fedora 26 with GNOME 3.24.2 and clipboard indicator with this PR works for me on Wayland. I can select Xorg or Wayland before open a session. So I can't investigate, sorry.

Maybe you can find a log on your computer.

**Comment by danboid** [comment]
Regular Ubuntu 17.10 defaults to using GNOME / Wayland now (instead of Unity / X.org) so its important GNOME extensions work under Wayland too.

I just tried logging into GNOME/Wayland again. The cursor shows but the panel / dock etc never show, Alt+F2 fails to open the GNOME shell command and I can't switch between VTs - I'm not sure Wayland supports that.

Surely there is a way to debug GNOME extensions that prevent GNOME shell starting without having to use another machine and SSH etc, @Tudmotu ?

**Comment by danboid** [comment]
$ journalctl /usr/bin/gnome-shell -f -o cat
Some code accessed the property 'SHELL_KEYBINDINGS_SCHEMA' on the module 'windowManager'. That property was defined with 'let' or 'const' inside the module. This was previously supported, but is not correct according to the ES6 standard. Any symbols to be exported from a module must be defined with 'var'. The property access will work as previously for the time being, but please fix your code anyway.
Some code accessed the property 'SettingsSchema' on the module 'prefs'. That property was defined with 'let' or 'const' inside the module. This was previously supported, but is not correct according to the ES6 standard. Any symbols to be exported from a module must be defined with 'var'. The property access will work as previously for the time being, but please fix your code anyway.
Some code accessed the property 'Fields' on the module 'prefs'. That property was defined with 'let' or 'const' inside the module. This was previously supported, but is not correct according to the ES6 standard. Any symbols to be exported from a module must be defined with 'var'. The property access will work as previously for the time being, but please fix your code anyway.
JS WARNING: [/home/dan/.local/share/gnome-shell/extensions/clipboard-indicator@tudmotu.com/extension.js 551]: assignment to undeclared variable CACHE_ONLY_FAVORITE
Failed to load background 'file:///usr/share/backgrounds/gnome/adwaita-lock.jpg': Error opening file /usr/share/backgrounds/gnome/adwaita-lock.jpg: No such file or directory
Error looking up permission: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.impl.portal.PermissionStore was not provided by any .service files
JS WARNING: [/home/dan/.local/share/gnome-shell/extensions/clipboard-indicator@tudmotu.com/extension.js 162]: reference to undefined property "favorite"
JS WARNING: [/home/dan/.local/share/gnome-shell/extensions/dash-to-panel@jderose9.github.com/taskbar.js 750]: reference to undefined property 0
Window manager warning: "XF86RFKill" is not a valid accelerator
GNOME Shell started at Wed Nov 08 2017 00:10:49 GMT+0000 (GMT)
JS WARNING: [/home/dan/.local/share/gnome-shell/extensions/dash-to-panel@jderose9.github.com/windowPreview.js 1001]: reference to undefined property 0
_cogl_buffer_fini: assertion '!(buffer->flags & COGL_BUFFER_FLAG_MAPPED)' failed
g_signal_handler_disconnect: assertion 'handler_id > 0' failed

**Comment by freeroot** [comment]
JS WARNING: [/home/dan/.local/share/gnome-shell/extensions/clipboard-indicator@tudmotu.com/extension.js 551]: assignment to undeclared variable CACHE_ONLY_FAVORITE

Try to add `let CACHE_ONLY_FAVORITE  = false;` in line 40 of extension.js
If it works, I will commit it.

**Comment by danboid** [comment]
I tried adding

let CACHE_ONLY_FAVORITE = false;

but Wayland GNOME is still freezing before the desktop loads.

$ journalctl /usr/bin/gnome-shell -f -o cat
Some code accessed the property 'Fields' on the module 'prefs'. That property was defined with 'let' or 'const' inside the module. This was previously supported, but is not correct according to the ES6 standard. Any symbols to be exported from a module must be defined with 'var'. The property access will work as previously for the time being, but please fix your code anyway.
JS WARNING: [/home/dan/.local/share/gnome-shell/extensions/clipboard-indicator@tudmotu.com/extension.js 551]: assignment to undeclared variable CACHE_ONLY_FAVORITE
JS WARNING: [resource:///org/gnome/shell/ui/workspaceThumbnail.js 892]: reference to undefined property "_switchWorkspaceNotifyId"
Failed to load background 'file:///usr/share/backgrounds/gnome/adwaita-lock.jpg': Error opening file /usr/share/backgrounds/gnome/adwaita-lock.jpg: No such file or directory
Error looking up permission: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.impl.portal.PermissionStore was not provided by any .service files
JS WARNING: [/home/dan/.local/share/gnome-shell/extensions/clipboard-indicator@tudmotu.com/extension.js 162]: reference to undefined property "favorite"
Window manager warning: "XF86RFKill" is not a valid accelerator
JS WARNING: [/home/dan/.local/share/gnome-shell/extensions/dash-to-panel@jderose9.github.com/taskbar.js 750]: reference to undefined property 0
GNOME Shell started at Wed Nov 08 2017 01:32:02 GMT+0000 (GMT)
JS WARNING: [/home/dan/.local/share/gnome-shell/extensions/dash-to-panel@jderose9.github.com/windowPreview.js 1001]: reference to undefined property 0
_cogl_buffer_fini: assertion '!(buffer->flags & COGL_BUFFER_FLAG_MAPPED)' failed
g_signal_handler_disconnect: assertion 'handler_id > 0' failed

**Comment by danboid** [comment]
Ooops!

Sorry! This PR is working under wayland actually. I had installed another new GNOME extension earlier (EasyScreenCast) which was working eaelier but was making wayland not start, not the clipboard plugin. Sorry about that!

It seems to be working as expected now - thanks!

**Comment by freeroot** [comment]
Good :)

For the other error : 

JS WARNING: [/home/dan/.local/share/gnome-shell/extensions/clipboard-indicator@tudmotu.com/extension.js 162]: reference to undefined property "favorite"

I don't understand because there is this test before `if (typeof buffer === 'string')` to avoid this error..

File: utils.js (hunk 3 above)
Comment (freeroot) [reply utils.js]: I put again this code but without deleting favorites. See 4149e96

**Comment by freeroot** [comment]
I think it's good to be merge now.

I have another commit to place favorites at the top of the menu (with a second one) but I can do another PR for it. This PR for favorite items and the other one for favorites' menu.

**Comment by Tudmotu** [comment]
Apologies this is taking so long to merge, but I need to test this a bit before merging and didn't find the time yet.

File: extension.js [hunk 4]
```diff
@@ -34,16 +34,16 @@ const prettyPrint = Utils.prettyPrint;
...
-let MAX_REGISTRY_LENGTH  = 15;
-let MAX_ENTRY_LENGTH     = 50;
-let CACHE_ONLY_FAVORITE  = false;
-let DELETE_ENABLED       = true;
-let ENABLE_KEYBINDING    = true;
-let PRIVATEMODE          = false;
-let NOTIFY_ON_COPY       = true;
-let MAX_TOPBAR_LENGTH    = 15;
-let TOPBAR_DISPLAY_MODE  = 1; //0 - only icon, 1 - only clipbord content, 2 - both
+var TIMEOUT_MS           = 1000;
```
Comment (Tudmotu) [inline extension.js]: Why did you change these?

Comment (freeroot) [reply extension.js]: In dandoid's log above, there are those lines : 

Some code accessed the property 'Fields' on the module 'prefs'. That property was defined with 'let' or 'const' inside the module. This was previously supported, but is not correct according to the ES6 standard. Any symbols to be exported from a module must be defined with 'var'. The property access will work as previously for the time being, but please fix your code anyway.

Comment (Tudmotu) [reply extension.js]: Ah, ok. But that warning is referring to the ``Fields`` variable from the ``prefs.js`` file. See it [here](https://github.com/Tudmotu/gnome-shell-extension-clipboard-indicator/blob/master/prefs.js#L12). The warning is actually about the [usage of that variable](https://github.com/Tudmotu/gnome-shell-extension-clipboard-indicator/blob/master/extension.js#L488) from within ``extension.js``.

To fix this warning, just change the ``const Fields = { ...`` in ``perfs.sj`` to ``var Fields = { ...``. Also, to make sure your changes do resolve the issue, you can take a look at the logs using journalctl: ``$ journalctl -f /usr/bin/gnome-shell``.

I am unsure how @danboid reproduced this issue, but I think changing the ``Fields`` variable definition should fix it.

Can you please change these back to ``let``?
Thanks!

Comment (Tudmotu) [reply extension.js]: You know what, never mind. I am testing it now anyway, so I will make the changes.
Again, thank you for this feature and the effort! :smiley:

**Comment by freeroot** [comment]
Is it ok ?

**Comment by Tudmotu** [comment]
Yeah, it's awesome. Gonna merge it, just need to make those small changes.

**Comment by freeroot** [comment]
What are the small changes that you've told about ?

I've came back to let for variables definitions, is it that ?

**Comment by Tudmotu** [comment]
Thank you for this awesome PR! :)
I will upload it to ego soon.
